pyamg/amg_core/ruge_stuben.h

```diff
@@ -956,345 +943,1339 @@ void cr_helper(const I A_rowptr[],  const int A_rowptr_size,
+                                        const I Sj[], const int Sj_size,
+                                        const T Sx[], const int Sx_size,
+                                        const I splitting[], const int splitting_size,
+                                        const I Pp[], const int Pp_size,
+                                              I Pj[], const int Pj_size,
+                                              T Px[], const int Px_size)
+{
+
+    for (I i = 0; i < n_nodes; i++) {
+        // If node i is a C-point, then set interpolation as injection
+        if(splitting[i] == C_NODE) {
+            Pj[Pp[i]] = i;
+            Px[Pp[i]] = 1;
+        } 
+        // Otherwise, use extended+i distance-two AMG interpolation formula
+        // (see Eqs. 4.10-4.11 in [0])
+        else {
+
+            // -------------------------------------------------------------------------------- //
+            // -------------------------------------------------------------------------------- //
+            // Calculate outer denominator
+            T denominator = 0.0;
+
+            // Start by summing entire row of A
+            for (I mm = Ap[i]; mm < Ap[i+1]; mm++) {
+                denominator += Ax[mm];
+            }
+
+            // Then subtract off the strong connections so that you are left with 
+            // denominator = a_ii + sum_{m in weak connections} a_im
+            for (I mm = Sp[i]; mm < Sp[i+1]; mm++) {
+                I this_point = Sj[mm];
+                if ( this_point != i ) {
+                    denominator -= Sx[mm]; // making sure to leave the diagonal entry in there
+                }
+                // Subtract off distance-two strong C connections that are also distance-one weak
+                // connections (i.e. not yet removed from the weak connections in denominator)
+                if (splitting[this_point] == F_NODE && (this_point != i)) {
+
+                    for (I ff = Sp[this_point]; ff < Sp[this_point+1]; ff++) {
+                        I d2_point = Sj[ff];
+
+                        // Strong C-connections to strong F-connections (distance two C connections)
+                        if (splitting[d2_point] == C_NODE) {
+
+                            // Add connection a_kl if present in matrix (search over kth row in A
+                            // for connection). Only add if sign of a_kl does not equal sign of a_kk
+                            for (I search_ind = Ap[i]; search_ind < Ap[i+1]; search_ind++) {
+                                if (Aj[search_ind] == d2_point) {
+                                    denominator -= Ax[search_ind];
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Sum over strongly connected fine points for outer denominator
```

Comment:
I also wonder if we can remove the duplicate code between ext and ext+i.  We could add a boolean flag "plusi", and then enclose the below `for (I kk =...)` loop inside an `if(plusi)` block. 

[ Note, I'm not adding any comments to the ext interp routine, in hopes they can be combined ]